### PR TITLE
fixed provider set delay NullPointerException

### PR DESCRIPTION
**Pull request by Leishunyu**
## What is the purpose of the change

https://github.com/apache/incubator-dubbo/issues/3952

fixed this bug

## Brief changelog

provider's delay has not set in serviceconfig 

## Verifying this change

set provider's delay in  serviceconfig 

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java [hunk 1]
```diff
@@ -360,6 +360,7 @@ private boolean shouldDelay() {
         if (delay == null && provider != null) {
             delay = provider.getDelay();
         }
+        this.delay = delay;
```
Comment (ralf0131) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java]: Hi, -1 to this change.

`shouldDelay` is a method that appears read-only. Modifying value within this method does not match with how this method should work.

I'd suggest to get the actual value of delay in `org.apache.dubbo.config.ServiceConfig#export`, or abstract it into a new method.

Comment (Leishunyu) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java]: ok , i change it

Comment (Leishunyu) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java]: i abstract it into a new method , can you help me see see

Comment (beiwei30) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java]: don't do this, the original change is good enough. The second change is unnecessary and make the code more complex.

Comment (beiwei30) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java]: pls. simply revert 95b394a6033e5a787e70fbdb1a3dd3b0dbc0aeb4. I will follow up.

**Comment by beiwei30** [comment]
LGTM. I will follow up to make it better, since I also need to take care *shouldExport* too. Thanks.

**Comment by ralf0131** [comment]
> LGTM. I will follow up to make it better, since I also need to take care _shouldExport_ too. Thanks.

This change does not look good to me. Why it is getting merged?
How would you planning to improve it?

**Comment by beiwei30** [comment]
@ralf0131 @Leishunyu pls. review https://github.com/apache/incubator-dubbo/pull/3959

**Comment by ralf0131** [comment]
The pull request looks good to me.
IMO, it would be better to let @Leishunyu to improve it.

**Pull request merged**